pkg/apis/application/v1alpha1/app_project_types.go

```diff
@@ -341,6 +343,44 @@ func (proj AppProject) IsResourcePermitted(groupKind schema.GroupKind, namespace
 	return true
 }
 
+// IsAppOfAppsPermitted returns whether a resource follows the app of apps rules of this project
+func (proj AppProject) IsAppOfAppsPermitted(un *unstructured.Unstructured) (string, bool) {
+	groupVersionKind := un.GroupVersionKind()
+	group := groupVersionKind.Group
+	kind := groupVersionKind.Kind
+	if group != application.Group || kind != application.ApplicationKind {
+		return "", true
+	}
+	project, _, err := unstructured.NestedString(un.Object, "spec", "project")
+	if err != nil {
+		panic(fmt.Errorf(
```

Comment:
Could this function return an error instead of panicking?

Reply:
I wasn't entirely sure what to do in this case, as the other related functions didn't seem to return an error. I was originally planning on modifying the IsResourcePermitted function and I wanted to avoid changing the API, and this is probably a leftover from when I split it out into its own function. If you'd prefer to return the error, then that's fine.

Reply:
I changed it to bubble the error back to the user, however, the e2e CI tests appear to be failing in a way that seems unrelated to the change. I managed to get it to run locally, but it fails on a different test due to a timeout, but the expected state appears to be reached after the test gives up. To confirm, I checked out master and ran e2e against it, and it failed on a different test. Are the e2e known to be flaky, or is something else going on?

Reply:
e2e is known to be flaky, and today might have to be the day I finally do something to fix it. 😄 Resterted the CI check.